explorer/interpreter/matching_impl_set.cpp

```diff
@@ -0,0 +1,163 @@
+  void VisitValue(const TypeType*) { Collect(ValueKey::TypeType); }
+
+  void VisitValue(const BoolType*) { Collect(ValueKey::BoolType); }
+
+  void VisitValue(const IntType*) { Collect(ValueKey::IntType); }
+
+  void VisitValue(const StringType*) { Collect(ValueKey::StringType); }
+
+  void VisitValue(const StaticArrayType* array) {
+    Collect(ValueKey::ArrayType);
+    Collect(&array->element_type());
+  }
+
+  void VisitValue(const PointerType* pointer) {
+    Collect(ValueKey::PointerType);
+    Collect(&pointer->pointee_type());
+  }
+
+  void VisitValue(const StructType* struct_type) {
+    for (auto [name, type] : struct_type->fields()) {
+      Collect(type);
+    }
+  }
+
+  void VisitValue(const NominalClassType* class_type) {
+    VisitDeclarationAndArgs(class_type->declaration(), class_type->bindings());
+  }
+
+  void VisitValue(const MixinPseudoType* mixin_type) {
+    VisitDeclarationAndArgs(mixin_type->declaration(), mixin_type->bindings());
+  }
+
+  void VisitValue(const InterfaceType* iface_type) {
+    VisitDeclarationAndArgs(iface_type->declaration(), iface_type->bindings());
+  }
+
+  void VisitValue(const NamedConstraintType* constraint_type) {
+    VisitDeclarationAndArgs(constraint_type->declaration(),
+                            constraint_type->bindings());
+  }
+
+  void VisitValue(const ChoiceType* choice_type) {
+    VisitDeclarationAndArgs(choice_type->declaration(),
+                            choice_type->bindings());
+  }
+
+  void VisitDeclarationAndArgs(const Declaration& declaration,
+                               const Bindings& bindings) {
+    Collect(match_->parent_->GetKeyForDeclaration(declaration));
+    for (auto [key, value] : bindings.args()) {
+      Collect(value);
+    }
+  }
+
+ private:
+  Match* match_;
+};
+
+auto MatchingImplSet::GetKeyForDeclaration(const Declaration& declaration)
+    -> ValueKey {
+  auto [it, added] = declaration_keys_.insert(
+      {&declaration, ValueKey(int(ValueKey::FirstDeclarationKey) +
+                              declaration_keys_.size())});
+  return it->second;
+}
+
+MatchingImplSet::Match::Match(Nonnull<MatchingImplSet*> parent,
+                              Nonnull<const ImplScope::Impl*> impl,
+                              Nonnull<const Value*> type,
+                              Nonnull<const Value*> interface)
+    : parent_(parent), impl_(impl), type_(type), interface_(interface) {
+  // Build our signature.
+  LeafCollector collector(this);
+  collector.Collect(type);
+  collector.Collect(interface);
+
+  parent_->matches_.push_back(this);
+}
+
+MatchingImplSet::Match::~Match() {
+  CARBON_CHECK(parent_->matches_.back() == this) << "match stack broken";
+  parent_->matches_.pop_back();
+}
+
+auto MatchingImplSet::Match::Check(SourceLocation source_loc)
+    -> ErrorOr<Success> {
+  for (auto* match : parent_->matches_) {
+    if (match != this && match->impl_ == impl_) {
+      bool all_greater_or_equal = true;
+      bool any_greater = false;
```

Comment:
I think it may help here to have some implementation comments about what comparisons you're tracking in these vars. i.e., "all_greater_or_equal" compared to what?

Perhaps the separate comment requests in the header would be another way to address that sufficiently; I'm mainly trying to give the feedback that I think more information would help, not necessarily where.

Reply:
Added comments in both places, let me know if it's too much now.